spack style: mypy checks installed environment packages

### Steps to reproduce

1. Create a managed environment with a view and some python sources, and install it
2. Run `spack style -t mypy`

### Error message

An example on my system:
```console
> spack style -t mypy                                                                                                                                                                                    17.5s
==> Running style checks on spack
  selected: mypy
==> Modified files
...
==> Running mypy checks
var/spack/environments/system/.spack-env/view/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2346: error: Positional-only parameters are only supported in Python 3.8 and greater  [syntax]
Found 1 error in 1 file (errors prevented further checking)
  mypy found errors
==> Error: spack style found errors
```

### Information on your system

* **Spack:** 0.21.0.dev0 (11bc27d984a1008f254abf6690c2d96c418b0c42)
* **Python:** 3.10.10
* **Platform:** darwin-ventura-m2
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -146,7 +146,23 @@ ignore_missing_imports = true
   # arguments, which don't work in 3.7, which causes mypy to bail out early if you have
   # numpy installed.
   [[tool.mypy.overrides]]
-  module = 'numpy'
+  module = [
+      'IPython',
+      'altgraph',
+      'attr',
+      'boto3',
+      'botocore',
+      'distro',
+      'jinja2',
+      'jsonschema',
+      'macholib',
+      'markupsafe',
+      'numpy',
+      'pyristent',
+      'pytest',
+      'ruamel.yaml',
+      'six',
+  ]
   follow_imports = 'skip'
   follow_imports_for_stubs = true
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -141,10 +141,11 @@ ignore_missing_imports = true
   ignore_errors = true
   ignore_missing_imports = true
 
-  # pytest (which we depend on) optionally imports numpy, which requires Python 3.8 in
-  # recent versions. mypy still imports its .pyi file, which has positional-only
-  # arguments, which don't work in 3.7, which causes mypy to bail out early if you have
-  # numpy installed.
+  # Spack imports a number of external packages, and they *may* require Python 3.8 or
+  # higher in recent versions. This can cause mypy to fail because we check for 3.7
+  # compatibility. We could restrict mypy to run for the oldest supported version (3.7),
+  # but that means most developers won't be able to run mypy, which means it'll fail
+  # more in CI. Instead, we exclude these imported packages from mypy checking.
   [[tool.mypy.overrides]]
   module = [
       'IPython',

